File: src/main/java/kr/co/bookand/backend/common/domain/dto/PageResponse.java
package kr.co.bookand.backend.common.domain.dto;

import lombok.Builder;
import lombok.Getter;
import org.springframework.data.domain.Page;

import java.util.List;

@Getter
@Builder
public class PageResponse<T> {
    private final int totalPages;
    private final long totalElements;
    private final boolean last;
    private final List<T> content;

    public static <T> PageResponse<T> of(Page<T> page) {
        return PageResponse.<T>builder()
                .totalPages(page.getTotalPages())
                .totalElements(page.getTotalElements())
                .last(page.isLast())
                .content(page.getContent())
                .build();
    }

    public static <T> PageResponse<T> ofCursor(Page<T> page, Long totalElements) {
        return PageResponse.<T>builder()
                .totalPages(page.getTotalPages())
                .totalElements(totalElements)
                .last(page.isLast())
                .content(page.getContent())
                .build();
    }

    public static long getTotalElements(long totalElements, Long cursorElements) {
        if (cursorElements != null) {
            totalElements += cursorElements;
        }
        return Math.max(totalElements, 0);
    }
}

File: src/main/kotlin/kr/co/bookand/backend/common/PageResponse.kt
package kr.co.bookand.backend.common

import org.springframework.data.domain.Page

data class PageResponse<T>(
    val totalPages: Int,
    val totalElements: Long,
    val last: Boolean,
    val content: List<T>
) {
    companion object {
        fun <T> of(page: Page<T>): PageResponse<T> {
            return PageResponse(
                totalPages = page.totalPages,
                totalElements = page.totalElements,
                last = page.isLast,
                content = page.content
            )
        }

        fun <T> ofCursor(page: Page<T>, totalElements: Long): PageResponse<T> {
            return PageResponse(
                totalPages = 0,
                totalElements = totalElements,
                last = page.isLast,
                content = page.content
            )
        }

        fun getTotalElements(totalElements: Long, cursorElements: Long?): Long {
            var updatedTotalElements = totalElements
            cursorElements?.let {
                updatedTotalElements += it
            }
            return updatedTotalElements.coerceAtLeast(0)
        }
    }
}